### PR TITLE
feat(a2asrv): add AuditLogger for structured A2A call auditing

### PR DESCRIPTION
**Pull request by kuangmi-bit**
## Summary

Add `a2asrv/auditlogger` — a server-side `CallInterceptor` that records structured audit events for every A2A protocol method invocation.

## Motivation

The existing `log/` package provides unstructured logging. Enterprise deployments need structured, machine-readable audit trails for compliance, monitoring, and debugging. The audit event schema aligns with IBM SMF Type 110 (CICS transaction audit): who called what method, when, with what result, and at what duration.

## Design

- **`AuditLogger`**: a `CallInterceptor`. `Before()` records start time; `After()` builds and writes the event.
- **`AuditWriter`**: pluggable interface. Ships with:
  - `InMemoryWriter`: for tests and dev
  - `JSONLWriter`: production-ready, concurrent-safe JSON lines file
- Events distinguish `success`, `error`, and `cancelled` results
- `CallContext.User.Name` used as agent identifier

## Example

```go
w, _ := auditlogger.NewJSONLWriter("/var/log/a2a/audit.jsonl")
al := auditlogger.NewAuditLogger(w)
handler := a2asrv.NewHandler(executor, a2asrv.WithCallInterceptors(al))
```

## Testing

7 tests pass: success/error/cancelled result types, multiple events, no-user edge case, JSONL file writer, concurrent writes (200 events across 10 goroutines).

### CODE REVIEW

File: a2asrv/costlimiter/costlimiter.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+
+// Reserve implements CostStore.
+func (s *InMemoryCostStore) Reserve(_ context.Context, scope string, cost int64) (bool, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	v, ok := s.budget[scope]
+	if !ok {
+		s.budget[scope] = -cost
+		return true, nil
+	}
```
[the remark this answers is not in the document]
Comment (kuangmi-bit) [reply a2asrv/costlimiter/costlimiter.go:205]: Done. `Reserve()` now returns `true, nil` for unlimited scopes without modifying the map — no more memory leak.

File: a2asrv/auditlogger/auditlogger.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+	ev.Tenant = callCtx.Tenant()
+
+	exts := callCtx.Extensions()
+	if exts != nil {
+		ev.Extensions = exts.RequestedURIs()
+	}
+
+	if resp != nil {
+		if resp.Err != nil {
+			if resp.Err == context.Canceled {
```
[the remark this answers is not in the document]
Comment (kuangmi-bit) [reply a2asrv/auditlogger/auditlogger.go]: Done. Changed to `errors.Is(resp.Err, context.Canceled)` for proper wrapped-error detection.

File: a2asrv/auditlogger/auditlogger.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	b, err := json.Marshal(ev)
+	if err != nil {
+		return err
+	}
+	b = append(b, '\n')
+	_, err = w.file.Write(b)
+	return err
+}
```
[the remark this answers is not in the document]
Comment (kuangmi-bit) [reply a2asrv/auditlogger/auditlogger.go:198]: Done. Moved `json.Marshal` outside the critical section — only `file.Write` holds the lock now.

File: a2asrv/costlimiter/costlimiter.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+			ErrBudgetExceeded, scope, cost, available)
+	}
+
+	ok, err := cl.store.Reserve(ctx, scope, cost)
+	if err != nil {
+		return ctx, nil, fmt.Errorf("costlimiter: reserving budget: %w", err)
+	}
+	if !ok {
+		return ctx, nil, fmt.Errorf("%w: scope %q: budget exhausted", ErrBudgetExceeded, scope)
+	}
```
[the remark this answers is not in the document]
Comment (kuangmi-bit) [reply a2asrv/costlimiter/costlimiter.go]: Done. Removed the redundant `Available()` check. `Reserve()` is already atomic — the check-then-act window is eliminated.

**Comment by nahapetyan-serob** [comment]
Thanks @kuangmi-bit.

The motivation isn't accurate. The existing log/ package is a wrapper around log/slog (Go's structured logging stdlib), not unstructured logging. And a2asrv.NewLoggingInterceptor (a2asrv/logging.go) is already a CallInterceptor that records method + duration + error on every A2A call using structured slog attributes, with a configurable level and optional payload capture. It covers the bulk of what this PR proposes.

**Comment by kuangmi-bit** [comment]
Thanks for the catch — you are right. The PR description called the output of `NewLoggingInterceptor` "unstructured," and that is inaccurate. `log/slog` is Go's structured logging stdlib, and the interceptor already emits method, duration, and error as structured attributes. I should not have framed it that way. The real motivation is not about structured vs. unstructured — it's that **audit and logging serve different consumers with different requirements**.

**Logging** (`NewLoggingInterceptor` → `slog`):
- Operational observability for developers and operators.
- Lives in the same sink as debug, info, and error logs — mixed with everything else.
- Schema is intentionally lightweight: method, duration, error.
- Format and destination are whatever the `slog` handler emits (text, JSON, etc.).

**Auditing** (the proposed `AuditLogger`):
- Compliance trail for SIEM / SMF / audit systems — a separate channel.
- Writes to a pluggable `AuditWriter` interface (JSONL file, Kafka, Splunk, etc.) — never mixed with operational logs.
- Richer schema: `agent_id`, `tenant`, structured `Result` enum (success / error / cancelled), `extensions`, `context_id` — things that matter for audit but would be noise in an application log.
- Immutable audit events aligned with SMF Type 110/80, suitable for compliance retention policies that are independent of log rotation.

The distinction matters when compliance asks "show me every A2A call, with tenant and agent identity, in a tamper-proof format, for the last 90 days" — that is a fundamentally different ask than "what is my server doing right now."

All that said — if the maintainers feel this does not warrant a separate package given the API surface concern (and for consistency with how #377 and #378 were handled), I am happy to close. Just wanted to clarify the actual motivation first, since the original description did a poor job of it. Appreciate the review.

**Comment by yarolegovich** [comment]
audit logging is unarguably important, but this implementation is very opinionated for a low-level SDK. it could be a good [a2a-samples](https://github.com/a2aproject/a2a-samples) contribution
